lib/core/src/restore_onchain.rs

```diff
@@ -0,0 +1,334 @@
+                                None
+                            }
+                            (_, Err(e)) => {
+                                error!("Failed to get swap script for Receive Swap {swap_id}: {e}");
+                                None
+                            }
+                        }
+                    })
+                    .collect();
+            let receive_swap_immutable_db_size = receive_swap_immutable_db.len();
+            info!("Receive Swap immutable DB: {receive_swap_immutable_db_size} rows");
+
+            Ok(ImmutableDb {
+                send_db: send_swap_immutable_db,
+                receive_db: receive_swap_immutable_db,
+            })
+        }
+    }
+}
+
+/// Methods to restore the swap tx IDs from the onchain data
+pub(crate) mod onchain {
+    use std::collections::HashMap;
+
+    use anyhow::{anyhow, ensure, Result};
+    use boltz_client::boltz::CreateReverseResponse;
+    use boltz_client::LBtcSwapScript;
+    use log::{error, info};
+    use lwk_wollet::elements::{Script, Txid};
+    use lwk_wollet::{History, WalletTx};
+
+    use crate::restore_onchain::immutable::ImmutableDb;
+    use crate::sdk::LiquidSdk;
+
+    /// A map of all our known onchain txs, indexed by tx ID
+    pub(crate) struct TxMap {
+        outgoing_tx_map: HashMap<Txid, WalletTx>,
+        incoming_tx_map: HashMap<Txid, WalletTx>,
+    }
+    impl TxMap {
+        pub(crate) fn from_raw_tx_map(raw_tx_map: HashMap<Txid, WalletTx>) -> Self {
+            let (outgoing_tx_map, incoming_tx_map): (
+                HashMap<Txid, WalletTx>,
+                HashMap<Txid, WalletTx>,
+            ) = raw_tx_map
+                .into_iter()
+                .partition(|(_, tx)| tx.balance.values().sum::<i64>() < 0);
+
+            Self {
+                outgoing_tx_map,
+                incoming_tx_map,
+            }
+        }
+    }
+
+    pub(crate) struct RecoveredOnchainDataSend {
+        lockup_tx_ids: HashMap<String, Txid>,
+        refund_tx_ids: HashMap<String, Txid>,
+    }
+
+    pub(crate) struct RecoveredOnchainDataReceive {
+        lockup_claim_tx_ids: HashMap<String, (Txid, Txid)>,
+    }
+
+    pub(crate) struct RecoveredOnchainData {
+        send: RecoveredOnchainDataSend,
+        receive: RecoveredOnchainDataReceive,
+    }
+
+    impl LiquidSdk {
+        pub(crate) async fn recover_from_onchain(&self, tx_map: TxMap) -> Result<()> {
+            let immutable_db = self.get_immutable_db_data().await?;
+
+            let _recovered = self.get_onchain_data(tx_map, immutable_db).await?;
+
+            // TODO Persist updated swaps
+            // TODO     Send updates with found txids
+            // TODO     How to set tx IDs without having to also set state?
+            // self.send_swap_state_handler
+            //     .update_swap_info(&swap_id, Failed, None, None, Some(&refund_tx_id_str))
+            //     .await?;
+
+            Ok(())
+        }
+        pub(crate) async fn get_onchain_data(
```

Comment:
The whole process contains lots of steps and since there are multiple types of data mapping structures  and lots of intermediate steps it makes it hard to follow the code.

In my mind the process could be simplify by following these 4 steps:
1. Fetch wallet transactions, map them by tx id.
2. Fetch two kind of swap items from db (receive, send), map them by script (e.g  Map<script, (swap_id, type)>
3. Fetch script history for all swaps together
4. For each script history:
- Find lockup and claim and add the info to the swap items.

The code roughly does that but I think it can be simplified.
For example we have:
`receive_db: HashMap<String, (CreateReverseResponse, LBtcSwapScript, Script)>`
and 
```
let mut receive_composite_data: HashMap<
                String,
                (CreateReverseResponse, LBtcSwapScript, Script, &Vec<History>)
```
Perhaps it is better and simpler to create a struct with the composite fields ( CreateReverseResponse, LBtcSwapScript, Script, &Vec<History> )
and fill them.
Also instead of &Vec<History> can we populate in that struct directly the lockup & claim ? It will save another processing (mapping) later.

Reply:
Done, added custom types and simplified the various types used.